src/dns.rs

```diff
@@ -0,0 +1,464 @@
+
+        self.resolve_name(name)
+    }
+
+    /// Should be only called from outside tokio. From inside tokio, Async version should be used.
+    fn resolve_name(&self, name: Name) -> Result<IntoIter<IpAddr>, io::Error> {
+        let future = self.construct_future(name);
+        // Will panic if called from inside tokio.
+        tokio::runtime::current_thread::block_on_all(future)
+    }
+
+    fn construct_future(&self, name: Name) -> ResolveFuture {
+        // try to parse as a regular IpAddr first
+        name.as_str()
+            .parse::<IpAddr>()
+            .ok()
+            .map(ResolveFuture::IpAddr)
+            // Else start resolution
+            .unwrap_or_else(|| match self.0.as_ref() {
+                Some(resolver) => ResolveFuture::Custom(resolver.lookup_ip(name.as_str()), name),
+                None => ResolveFuture::System(name),
+            })
+    }
+}
+
+impl Resolve for DnsResolver {
+    type Addrs = IntoIter<IpAddr>;
+    /// A Future of the resolved set of addresses.
+    type Future = ResolveFuture;
+    /// Resolve a hostname.
+    fn resolve(&self, name: Name) -> Self::Future {
+        self.construct_future(name)
+    }
+}
+
+pub enum ResolveFuture {
+    /// Resolved
+    IpAddr(IpAddr),
+    /// Resolving using custom DNS servers
+    Custom(BackgroundLookupIp, Name),
+    /// Resolve using system resolver servers
+    System(Name),
+}
+
+impl ResolveFuture {
+    fn system_resolve(name: &Name) -> Poll<IntoIter<IpAddr>, io::Error> {
+        match (name.as_str(), 0).to_socket_addrs() {
+            Ok(ips) => Poll::Ok(Async::Ready(
+                ips.map(|socket_addr| socket_addr.ip())
+                    .collect::<Vec<_>>()
+                    .into_iter(),
+            )),
+            Err(error) => Poll::Err(error),
+        }
+    }
+}
+
+impl Future for ResolveFuture {
+    type Item = IntoIter<IpAddr>;
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        match self {
+            ResolveFuture::IpAddr(addr) => Poll::Ok(Async::Ready(vec![addr.clone()].into_iter())),
+            ResolveFuture::Custom(future, name) => match future.poll() {
+                Poll::Ok(Async::NotReady) => Poll::Ok(Async::NotReady),
+                Poll::Ok(Async::Ready(ips)) => {
+                    Poll::Ok(Async::Ready(ips.iter().collect::<Vec<_>>().into_iter()))
+                }
+                Poll::Err(error) => {
+                    match error.kind() {
+                        ResolveErrorKind::NoRecordsFound { .. } => debug!("No records found"),
+                        _ => {
+                            error!(message = "Error while resolving Domain Name", error = ?error);
+                        }
+                    }
+                    ResolveFuture::system_resolve(name)
+                }
+            },
+            ResolveFuture::System(name) => ResolveFuture::system_resolve(name),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
this mocking of the DNS resolution is something we have some integration test support for in the trust-dns integration-tests crate (which is unpublished). Should we look at publishing that crate to crates.io to make this type of testing easier?

Reply:
That would help.

Reply:
@bluejeklyll we would be happy to help contribute code as well! let us know what we can do to get that moving.

Reply:
For reference, here's the MockClient that I built for integration tests: https://github.com/bluejekyll/trust-dns/blob/master/tests/integration-tests/src/mock_client.rs

At the root of that lib are a few other types for forcing certain scenarios during testing: https://github.com/bluejekyll/trust-dns/blob/master/tests/integration-tests/src/lib.rs

This sets up in-memory authorities like you have in these tests: https://github.com/bluejekyll/trust-dns/blob/master/tests/integration-tests/src/authority.rs

Then here's an example of how I use all of that to test specific issues, like distrust-nxdomain :) https://github.com/bluejekyll/trust-dns/blob/master/tests/integration-tests/tests/name_server_pool_tests.rs#L308

So I think there's some overlap, and if you all want we could try and clean those up to be more useful outside the trust-dns libraries and be made available for use in cases like you have here.